Configurable lliveness probe method

charlcater opened the pull request:
## Description

Add the option to set a command for a liveness probe and fall back to the httpGet method if not set.

## Types of changes

_Put an `x` in the boxes that apply._

- [ ] Bugfix (`fix`): A non-breaking change which fixes an issue or bug
- [x] New Feature (`feat`): A non-breaking change which adds a new feature or new functionality
- [ ] Tests (`test`): Adding missing tests or correcting existing tests
- [ ] Refactoring (`refactor`): A code change that neither fixes a bug nor adds a feature, e.g Deleting unused or redundant code
- [ ] Documentation (`docs`): Documentation only changes
- [ ] Other (`style`, `build`, `ci`, `performance`): Formatting, linting or styling. Changes to the CI pipeline. Performance improvements. Changes that affect the build system.

## Further Information

Any relevant logs, screenshots, testing output, etc...

deontpearson reviewed (commented):
1. We should do the same to readiness probe too
2. We should make this a major version bump and rather change the way the liveness and readiness probe is defined. For example if `livenessProbe.http` then add the http stuff, or `if livenessProbe.exec` then add the exec stuff. Might need to think what happens if both are defined.

charlcater commented:
Thanks Deon. I agree with what you suggest. I've built this to be kind-of backwards compatible, but happy to make a major update if you think it is warranted

deontpearson commented on charts/aruba-uxi/values.example.yaml:
```diff
@@ -86,10 +86,15 @@ aruba-uxi:
       ## Enables the liveness probe to determine if the pod started up correctly (optional, default: {})
       livenessProbe:
-        ## Toggles the liveness probe (optional, default: true)
-        enabled: true
```
I think enabled should stay. It controls whether the whole liveness or readiness probe is added. Its different to execmethod or httpmethod

deontpearson commented on charts/aruba-uxi/templates/deployment.yaml:
```diff
@@ -109,7 +109,17 @@ spec:
           containerPort: {{ $data.port }}
           protocol: TCP
         {{- end }}
-        {{- if $readinessProbe.enabled }}
```
I think this should stay to easily toggle it on or off

deontpearson commented on charts/aruba-uxi/values.example.yaml:
```diff
@@ -86,10 +86,15 @@ aruba-uxi:
-        ## The API path to query for liveness testing (optional, default: /health/livez)
-        path: /health/livez
+        ## Toggles the command-type probe (optional, default: true)
+        execmethod: true
```
What do you think about labeling these values the same as the values added in the deployment. E.g

```yaml
livenessProbe:
    exec:
        enabled: true | false
        command: ...
    httpGet:
        enabled: true | false
        path: ...

```

deontpearson commented on charts/aruba-uxi/values.example.yaml:
```diff
@@ -86,10 +86,15 @@ aruba-uxi:
-        path: /health/livez
+        ## Toggles the command-type probe (optional, default: true)
+        execmethod: true
+        ## The command to use for liveness testing (optional, default: "")
+        ## Calling bash with the -c flag allows injection of the $HOSTNAME variable
```
If the command is optional I suggest making the default value something. E.g `bash -c uptime`. This way if the user enables the exec method and does not add a command they will not run into an error.

deontpearson replied:
```suggestion
        ## The command to use for liveness testing (optional, default: "")
        ## Calling bash with the -c flag allows injection of environment variables
```

deontpearson commented on charts/aruba-uxi/values.example.yaml:
```diff
@@ -86,10 +86,15 @@ aruba-uxi:
+        execmethod: true
+        ## The command to use for liveness testing (optional, default: "")
+        ## Calling bash with the -c flag allows injection of the $HOSTNAME variable
+        command: [
+          "bash",
+          "-c",
+          "celery --app=server.celery.tasks inspect ping -d celery@$HOSTNAME"
+        ]
```
Do you think its better to add a generic command here. For someone else reading this they might think it's only possible to to run a celery command. Maybe a command like `bash -c uptime`.  With the description stating that the command can be anything appropriate to determine the liveness of the application.

deontpearson commented on charts/aruba-uxi/values.example.yaml:
```diff
@@ -102,8 +107,8 @@ aruba-uxi:
       readinessProbe:
-        ## Toggles the readiness probe (optional, default: true)
-        enabled: true
+        ## Toggles the httpGet type probe (optional, default: true)
+        httpmethod: true
```
Same comment as above regarding the structure of this value

deontpearson reviewed (commented):
I made some comments inline. Do you think there could be some way to reduce the duplication of these lines

```yaml
          initialDelaySeconds: {{ default "10" $livenessProbe.initialDelaySeconds }}
          periodSeconds: {{ default "10" $livenessProbe.periodSeconds }}
          timeoutSeconds: {{ default "1" $livenessProbe.timeoutSeconds }}
          successThreshold: {{ default "1" $livenessProbe.successThreshold }}
          failureThreshold: {{ default "3" $livenessProbe.failureThreshold }}
```

charlcater commented:
OK, I addressed all your comments. I think that your suggestions keeps it closer in structure to what went before 👍 
From testing the only smallish caveat seems to be that both the `exec` and `httpGet` methods must be explicitly set to true / false if the probe is enabled, otherwise the output file fails to render

deontpearson commented on charts/aruba-uxi/values.example.yaml:
```diff
@@ -87,9 +87,24 @@ aruba-uxi:
+        ## Section to specify the httpGet-type liveness probe
+        httpGet:
+          ## Toggles the httpGet-type probe (optional, default: true)
+          enabled: false
```
```suggestion
          enabled: false
           ## The API path to query for readiness testing (optional, default: /health/livez)
          path: /health/livez
```
